### ISSUE TITLE
UnimplementedOp: DepthToSpace

### ISSUE THREAD

**Issue by ro99** [sonos/tract#554]
Hi! I am trying to run a model, that is based on UNet, and I am receiving the following error during into_optimized operation:

`Error: Translating node #321 "408" Unimplemented(DepthToSpace) ToTypedTranslator
`

I understand that there are some ONNX operations that are not supported yet. Is there anything that I can do to support this particular operation? I am willing to contribute, but I would need some directions first 😅.

**Comment by ro99** [comment]
I assumed that this would be similar to the `reshape.rs`, so I thought about something like this:

`core/src/ops/array/depth_to_space.rs`:

```rust
use crate::internal::*;

#[derive(Debug, Clone, PartialEq, Hash)]
pub enum DepthToSpaceMode {
    DCR,
    CRD,
}

impl Default for DepthToSpaceMode {
    fn default() -> DepthToSpaceMode {
        DepthToSpaceMode::DCR
    }
}

#[derive(Debug, Clone, new, Default, Hash)]
pub struct DepthToSpace {
    pub block_size: usize,
    pub mode: DepthToSpaceMode,
}

impl_dyn_hash!(DepthToSpace);

impl Op for DepthToSpace {
    fn name(&self) -> Cow<str> {
        "DepthToSpace".into()
    }
    op_core!();
    op_as_typed_op!();
}

impl EvalOp for DepthToSpace {
    fn is_stateless(&self) -> bool {
        true
    }

    fn eval(&self, mut inputs: TVec<Arc<Tensor>>) -> TractResult<TVec<Arc<Tensor>>> {
        let input = args_1!(inputs);
        let mut tensor = input.into_tensor();
        let shape: TVec<usize> = tensor.shape().into();

        let shape = match self.mode {
            DepthToSpaceMode::DCR => {
                tvec!(
                    shape[0],
                    shape[1] / (self.block_size * self.block_size),
                    shape[2] * self.block_size,
                    shape[3] * self.block_size
                )
            }
            DepthToSpaceMode::CRD => {
                tvec!(
                    shape[0],
                    shape[1] * self.block_size,
                    shape[2] * self.block_size,
                    shape[3] / (self.block_size * self.block_size),
                )
            }
        };
        unsafe {
            tensor.set_shape_unchecked(&*shape);
        }
        Ok(tvec!(tensor.into_arc_tensor()))
    }
}

impl TypedOp for DepthToSpace {
    fn output_facts(&self, inputs: &[&TypedFact]) -> TractResult<TVec<TypedFact>> {
        let mut fact = inputs[0].clone();
        match self.mode {
            DepthToSpaceMode::DCR => {
                fact.shape.set(1, fact.shape[1].clone() / (self.block_size * self.block_size));
                fact.shape.set(2, fact.shape[2].clone() * self.block_size);
                fact.shape.set(3, fact.shape[3].clone() * self.block_size);
            }
            DepthToSpaceMode::CRD => {
                fact.shape.set(1, fact.shape[1].clone() * self.block_size);
                fact.shape.set(2, fact.shape[2].clone() * self.block_size);
                fact.shape.set(3, fact.shape[3].clone() / (self.block_size * self.block_size));
            }
        };
        Ok(tvec!(fact))
    }

    as_op!();
}
```
But to be honest, I'm not sure if I'm doing this correctly... I don't even know if this should be on the core, maybe this should be only under onnx?

Talking about the onnx module, I'm still trying to figure out how to implement for it, and what are the other changes that need to be done (still learning about the framework, there are many points that are not clear for me 😅). 

Please let me know your thoughts on this. Any help/suggestion is appreciated!

Thanks!

**Comment by kali** [comment]
Hey, thanks for your interest in tract!

As far as I can tell, DepthToSpace and SpaceToDepth are fancy axis permuations. As such, they are high level operations: we only want to translate them to tract-core. We can put them in a new module under `onnx/src/ops/s2d.rs` maybe.

I suggest you try implementing them as implementation of the trait `Expansion`from `hir/src/ops/expandable.rs`. This way you basically have two big methods to implement: `rules` which is consume by the shape inference system, and `wire` to re-express the operators in terms of tract-core operators. You will find quite a few examples of Expansion in hir, onnx, and tensorflow if you need inspiration, specifically for the rules() method, the shape expression DSL being a bit tough to use sometimes. Don't forget to plug your op or ops into the register_all_ops tree to make them know to the onnx parser.

Then comes the wire method: I think you can achieve whan you need by a combination of a handful of AxisOp matching the numpy code given in the doc. AxisOp::Reshape just "cast" the shape, while AxisOp::Move allows to implement the permutation.

You can test the onnx samples with `cargo test -- --ignored depth` in `harness/onnx-test-suite`.

**Comment by kali** [comment]
Mmmm.... not really. What it will look like is a sequence of calls on `wire_node()` plugging in a sequence of various AxisOp (as I said before, I expect to see some Reshapes and Moves). `to_axis_op` is relatively complex in that it retrofits the too general Reshape from the training framework to a simple sequence of AxisOp (they will then get wired in the wire method for the general Reshape a few lines above).

A simpler example can be found in `hir/src/ops/array/rm_dims.rs`. But you will need more then one AxisOp to re-express the op: first some Reshapes to make the block axes appear, then 2 Moves to put them in the right place, and finally one or more Reshape to collapse them.

**Comment by ro99** [comment]
Thank you Kali, I believe I have [finished](https://gist.github.com/ro99/7aec88d716c0039a2082325d2ed621cb), just looking how to test now 😁.

**Comment by ro99** [comment]
Thanks for the inputs Kali, I just created a new PR for this.

**Comment by ro99** [comment]
Thanks for the inputs, I have finished and now I am doing some tests with my model. So far I observed that for the case with fixed input shape it is working: e.g.:

```rust
let model = tract_onnx::onnx()
        .model_for_path(MODEL_PATH)?
        .with_input_fact(
            0,
            InferenceFact::dt_shape(
                f32::datum_type(),
                &[1,3,256,256],
            ),
        )?
        .with_output_fact(0, InferenceFact::default())?
        .into_optimized()?
        .into_runnable()?;
```

But when I try with dynamic input shape it is not working.

```rust
    let model = tract_onnx::onnx()
        .model_for_path(MODEL_PATH)?
        .with_input_fact(
            0,
            InferenceFact::dt_shape(
                f32::datum_type(),
                &[
                    1usize.to_dim(),
                    3usize.into(),
                    height_symbol.into(),
                    width_symbol.into(),
                ],
            ),
        )?
        .with_output_fact(0, InferenceFact::default())?
        .into_optimized()?
        .into_runnable()?;
```

I receive the following error:

```
Error: Failed analyse for node #369 "410" InferenceConcat

Caused by:
    0: Infering facts
    1: Applying rule inputs[0].shape[2] == inputs[1].shape[2]
    2: Impossible to unify MulInt(2, Div(Add([Sym(Symbol('h', 0)), Val(31)]), 32)) with Div(Add([Sym(Symbol('h', 0)), Val(15)]), 16).
```
I am checking now if I am doing anything wrong with the arithmetic operations on TDim (at `compute_shape`), or maybe something in the rules... 

I have created a new branch [here](https://github.com/ro99/tract/tree/DepthToSpace-op).

**Comment by ro99** [comment]
Hmm, this is a little odd. Just to give some context of my model, the output from `DepthToSpace` operation will connect with a `Concat` operation (input[0]). The other input of `Concat` (input[1]) comes from a `BatchNormalization` operation... The error happens because these two inputs are supposed to be the same.

In the **working example** the input shape is fixed `(1,3,256,256)`, we have the following shape values at `DepthToSpace`:

```
inputs[0].shape -> [Val(1), Val(1024), Val(8), Val(8)]
outputs[0].shape -> [Val(1), Val(256), Val(16), Val(16)]

computed output shape -> [Val(1), Val(256), Val(16), Val(16)].   ...ok
```
In the `Concat` we have the following inputs shapes:

```
inputs[0].shape -> [Val(1), Val(256), Val(16), Val(16)]
inputs[1].shape -> [Val(1), Val(256), Val(16), Val(16)]
```

Here both inputs of concat are the same, so all good.

Now to the **not working example**. The input shape is dynamic `(1,3,h,w)`. We have the following shape values at `DepthToSpace`:

```
inputs[0].shape -------> [Val(1), Val(1024),          Div(Add([Sym(Symbol('h', 0)), Val(31)]), 32),         Div(Add([Sym(Symbol('w', 1)), Val(31)]), 32)]
outputs[0].shape ------> [Val(1), Val(256), MulInt(2, Div(Add([Sym(Symbol('h', 0)), Val(31)]), 32)), MulInt(2, Div(Add([Sym(Symbol('w', 1)), Val(31)]), 32))]

computed output shape -> [Val(1), Val(256), MulInt(2, Div(Add([Sym(Symbol('h', 0)), Val(31)]), 32)), MulInt(2, Div(Add([Sym(Symbol('w', 1)), Val(31)]), 32))].   ...ok
```

Ok, so the input and the output are the same, with the exception that for the width and height the values are multiplied by 2 (blocksize) on the output. This is the expected behavior from `DepthToSpace` operation. So far so good.

Now to the issue. In the `Concat` we have the following input values:

```
inputs[0].shape ---> [Val(1), Val(256), MulInt(2, Div(Add([Sym(Symbol('h', 0)), Val(31)]), 32)), MulInt(2, Div(Add([Sym(Symbol('w', 1)), Val(31)]), 32))]
inputs[1].shape ----> [Val(1), Val(256), Div(Add([Sym(Symbol('h', 0)), Val(15)]), 16), Div(Add([Sym(Symbol('w', 1)), Val(15)]), 16)]
```

Here these two are different, and so the error happens. 

But thinking from the arithmetic point of view, I understand that they are the same (here maybe I am interpreting something wrong). For instance:

**For the input[0] we have:**

Something ('h') that will divide by 32 and then the result is multiplied by 2. So for instance, if 'h' is 256 then:

`(256/32) * 2 == 16`

**For the input[1] we have:**

Something ('h') that will divide by 16. So for instance, if 'h' is 256 then:

`(256/16) == 16`

I may be interpreting the annotation given by the TDim incorrectly. Or maybe the MulInt(2)  should be applied (somehow) on `DepthToSpace`, I just do not know how. Example, `MulInt(1, Div(Add([Sym(Symbol('h', 0)), Val(15)]), 16))`

Any idea what could be wrong?

**Comment by ro99** [comment]
With 32h and 32w it is working !

**Comment by kali** [comment]
Good ! And your model makes no use of SpaceToDepth ?
